Ability to set pod environment variables on cluster resource

dmvolod opened the pull request:
Fixes #1566
Closes #1209
Closes #1422
Closes #907
Closes #1807

Please pay attention, that this PR implementing an individual `[]EnvVar` option to the cluster CR and has higher priority to the cluster `pod_environment_secret` and `pod_environment_configmap` options.

TODO (after approach approve)

- [x] Doc changes
- [x] Unit testing

dmvolod commented:
@FxKu could you please to have a look at the fix approach. If it will be suitable, I will add docs and unit-test for it. Thanks in advance.

cdmikechen commented:
Is it possible to add only some environment variables to pod? for example
```yaml
spec:
  env:
    - name: ENV_1
      value: VALUE_1
```

dmvolod commented:
> In CRD, Is there possible to add only some environment variables to pod? for example

Not sure, that's good approach as will need to support both setup for predefined ConfigMap/Secret and current one. Common way much better and didn't overload CR if it's not needed

dmvolod commented:
@FxKu could you please to find a time and have a look at the fix approach. If it will be suitable, I will add docs and unit-test for it. Thanks in advance.

FxKu commented:
Hey @dmvolod, while it looks elegant at first, I don't think it should be solved like this because only one way is possible - global config map or the need to add config maps for each cluster. The latter will be tedious if you need cluster-specific variables only for a handful of clusters while the rest of your fleet should be configured in the same way. Better have env variables from cluster-specific config maps / secrets to overwrite global ones - like we do it with other options, too. Probably, partly overwriting (in case of same env variables) will be most convenient.

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -941,6 +941,11 @@ func (c *Cluster) getPodEnvironmentSecretVariables() ([]v1.EnvVar, error) {
 
+func (c *Cluster) podEnvironmentObjectNameTemplate(template string) string {
+	objectNameTemplate := config.StringTemplate(template)
+	return objectNameTemplate.Format("cluster", c.clusterName().Name, "tprkind", acidv1.PostgresCRDResourceKind, "tprgroup", acidzalando.GroupName)
```
why should the template contain the CRD kind and group?

dmvolod replied:
I got the same template as operator has for the secret names to be unique name across different types and be a separating. Probably, this options can be omitted if not needed. Also, we could probably add a `CONTROLLER_ID` to have unique name across clusters and controllers.

dmvolod commented:
> Hey @dmvolod, while it looks elegant at first, I don't think it should be solved like this because only one way is possible - global config map or the need to add config maps for each cluster. The latter will be tedious if you need cluster-specific variables only for a handful of clusters while the rest of your fleet should be configured in the same way. Better have env variables from cluster-specific config maps / secrets to overwrite global ones - like we do it with other options, too. Probably, partly overwriting (in case of same env variables) will be most convenient.

Thanks for feedback, @FxKu . I'm also agree with approach to have a ConfigMap/Secret parameter for each cluster individually if needed and present. If you agree with this approach I will add these options to the CRD and implement logic to overwrite global parameters if CR parameters are present and find in the cluster ConfigMap/Secret.

cdmikechen commented:
I‘m looking forward to the progress of this PR. 
I have similar needs at present. We need to deploy multiple Postgres to different namespaces and back up to different minio services.

FxKu commented:
@cdmikechen note that you can run multiple operators next to each other configured differently. They could either watch only one namespace or you define a [CONTROLLER_ID]( https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#operators-with-defined-ownership-of-certain-postgres-clusters) in deployment and manifests to specify ownership over clusters.

cdmikechen commented:
> @cdmikechen note that you can already run multiple operators next to each other configured differently. They could either watch only one namespace or you define a [CONTROLLER_ID](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#operators-with-defined-ownership-of-certain-postgres-clusters) in deployment and manifests to specify ownership over clusters.

@FxKu
Thanks~ It looks like a good solution for my needs. I used to deploy only one operator to manage all namespaces, which will really bring me some trouble. 
I'll try it later and see if there will be any other problems.

dmvolod commented:
@FxKu please have a look at the new approach with individual `[]EnvVar` option again and has a higher priority to the global cluster options. Thanks in advance.

gc-jro commented:
Hi @dmvolod, thanks for this PR.

Do you still plan to add the possibility to refer to a cluster specific secret for environment variables in the postgresql custom resource?

I’ve a slightly different set of requirements, where we maintain our cluster configuration in Git, but aren’t allowed to store confidential information (like wal-g backup settings) in plain text. To circumvent this, we use SealedSecrets [1] to store the settings encrypted in Git. In our k8s cluster the SealedSecrets get decrypted into normal secrets that can be used by the cluster. I’ve also the requirement that we’ve multiple postgresql clusters in the same namespace, and the need that they use different wal-g backup settings.

Therefore, it would be great if in addition to use individual `[]EnvVar` settings in the postgresql CR, there would be an option to specify an individual secret in the postgresql CR that would be used for populating environment variables.
If you aren’t planning to add this option that’s totally fine, then I’ll ask my boss to get some time to patch the functionality in myself.

[1] https://github.com/bitnami-labs/sealed-secrets

dmvolod commented:
> 

Hi @gc-jro !

Yes, sure. Could you please to have a look at the `EnvVar` struct which is including `EnvVarSource` field and supports referencing to the Secret.
https://github.com/kubernetes/api/blob/master/core/v1/types.go#L1977-L2015

What do you think about it?

gc-jro commented:
Ah I see. Thanks a lot for clearing that up for me. I really like it, it'll solve all my problems :smile:

I'll just have to define something like
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: wal-g-envs
  namespace: yyy
data:
  BACKUP_NUM_TO_RETAIN: xxx
  ...
---
apiVersion: "acid.zalan.do/v1"
kind: postgresql
metadata:
  name: wal01-postgres-cluster
  namespace: yyy
spec:
  teamId: "wal01"
  volume:
    size: 20Gi
  env:
    - name: BACKUP_NUM_TO_RETAIN
      valueFrom:
        secretKeyRef:
          key: BACKUP_NUM_TO_RETAIN
          name: wal-g-envs
  ...
```
to forward the values from the secret to the database. Very neat.

Thank you very much. I can't wait to give it a spin once its merged into master.

FxKu commented:
Thank @dmvolod ! Like the new approach much better. As you already pointed out, docs and unit test are the remaining bits to get this merged.

sdudoladov commented on pkg/cluster/k8sres.go:
```diff
@@ -1000,20 +1010,31 @@ func (c *Cluster) getPodEnvironmentSecretVariables() ([]v1.EnvVar, error) {
 	return secretPodEnvVarsList, nil
 }
 
+func isEnvVarExist(envs []v1.EnvVar, key string) bool {
```
Minor naming thing: `is` would read better with an adjective, for instance `isEnvVarPresent`

dmvolod replied:
Good point, thanks, will change

sdudoladov commented on pkg/cluster/k8sres.go:
```diff
@@ -969,6 +975,10 @@ func (c *Cluster) getPodEnvironmentConfigMapVariables() ([]v1.EnvVar, error) {
 func (c *Cluster) getPodEnvironmentSecretVariables() ([]v1.EnvVar, error) {
 	secretPodEnvVarsList := make([]v1.EnvVar, 0)
 
+	if len(c.Spec.Env) > 0 {
+		secretPodEnvVarsList = append(secretPodEnvVarsList, c.Spec.Env...)
+	}
+
 	if c.OpConfig.PodEnvironmentSecret == "" {
 		return secretPodEnvVarsList, nil
 	}
```
What is the expected outcome of  this piece of code when one has in the manifest
```yaml
spec:
  env:
...
  - name: minio_secret_key
      valueFrom:
        secretKeyRef:
          name: my-custom-secret
          key: minio_secret_key
```
but ` c.OpConfig.PodEnvironmentSecret` is not set ?

dmvolod replied:
This should return `[]v1.EnvVar` from the Postgresql CR. i.e. 
```
  - name: minio_secret_key
      valueFrom:
        secretKeyRef:
          name: my-custom-secret
          key: minio_secret_key
```

FxKu commented on manifests/postgresql.crd.yaml:
```diff
@@ -194,6 +194,12 @@ spec:
+                nullable: true
+                items:
+                  type: object
+                  x-kubernetes-preserve-unknown-fields: true
```
please, add it also in the [chart](https://github.com/zalando/postgres-operator/blob/master/charts/postgres-operator/crds/postgresqls.yaml).

dmvolod replied:
Yes, thanks, will do

FxKu commented on docs/administrator.md:
```diff
@@ -706,6 +706,35 @@ data:
 
+### For individual cluster
+
+It is possible to set environment variables in the Spilo pods directly as parameters or
+as a link to secrets for each individual cluster, in addition to or instead of the
+global parameters given in the examples above.
```
```suggestion
It is possible to define environment variables directly in the Postgres cluster
manifest to configure it individually. The variables must be listed under the
`env` section in the same way you would do for [containers](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/).
Global parameters served from a custom config map or secret will be overridden.
```

FxKu commented on docs/administrator.md:
```diff
@@ -706,6 +706,35 @@ data:
 
+### For individual cluster
+
+It is possible to set environment variables in the Spilo pods directly as parameters or
+as a link to secrets for each individual cluster, in addition to or instead of the
+global parameters given in the examples above.
+
+In order to do this, you need to configure the cluster parameters, as in the example below
```
```suggestion
It is possible to define environment variables directly in the Postgres cluster
manifest to configure it individually. The variables must be listed under the
`env` section in the same way you would do for [containers](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/).
Global parameters served from a custom config map or secret will be overridden.
```

FxKu commented on docs/administrator.md:
````diff
@@ -706,6 +706,35 @@ data:
+          key: minio_secret_key
+```
+
+If global parameters are set at the level of the entire controller in the form of
+a ConfigMap or a Secret, then the individual cluster parameters will have a higher
+priority and the same keys will be overridden, and different ones will be merged into
+a single array of environment variable parameter values.
````
I think, this paragraph is obsolete after my suggestion from above.

dmvolod replied:
Ok. good point, done

FxKu commented:
I feel like this needs more unit tests. As per [docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#custom-pod-environment-variables) we currently say:
"In general the Operator will give preference to the globally configured variables, to not have the custom ones interfere with core functionality. Variables with the 'WAL_' and 'LOG_' prefix can be overwritten though, to allow backup and log shipping to be specified differently."

Might be worth to update this paragraph, too. Here's the corresponding [code](https://github.com/zalando/postgres-operator/blob/0dc370f15d5fd348937dd4eb631852e55df33e57/pkg/cluster/k8sres.go#L877), but I find it also weird that we keep appending. There's no `isEnvVarPresent` check happening. So another good test case.

Overall probably a bit confusing. Variables under `spec.env` override the global `pod_environment_configmap/secret` and partly override the global config, but then some global variables take precedence over individual cluster variables.

dmvolod commented:
> I feel like this needs more unit tests. As per [docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#custom-pod-environment-variables) we currently say: "In general the Operator will give preference to the globally configured variables, to not have the custom ones interfere with core functionality. Variables with the 'WAL_' and 'LOG_' prefix can be overwritten though, to allow backup and log shipping to be specified differently."
> 
> Might be worth to update this paragraph, too. Here's the corresponding [code](https://github.com/zalando/postgres-operator/blob/0dc370f15d5fd348937dd4eb631852e55df33e57/pkg/cluster/k8sres.go#L877), but I find it also weird that we keep appending. There's no `isEnvVarPresent` check happening. So another good test case.
> 
> Overall probably a bit confusing. Variables under `spec.env` override the global `pod_environment_configmap/secret` and partly override the global config, but then some global variables take precedence over individual cluster variables.

Yes, very good point. In this case, what do you think to move cluster `v1.EnvVar` logic later after the `getPodEnvironmentConfigMapVariables` and `getPodEnvironmentSecretVariables` calls and got a higher priority for the cluster parameters. Or we could keep logic "as is" and just adding more tests and descriptions?

dmvolod commented:
> I see that we even have a function to take care of duplicate env vars - so we might not even need the `isEnvVarPresent` check. I have [extended](https://gist.github.com/FxKu/0b4e0641f1c846a2a0edbc2c6ee6a0cc) `TestPodEnvironmentConfigMapVariables` with the behavior that reflects the docs. But's it's failing. Guess `deduplicateEnvVars` is doing something wrong then.

Thanks for pointing me. Will fix it and add an additional tests.

FxKu commented:
I got it working now. See my gist: https://gist.github.com/FxKu/0b4e0641f1c846a2a0edbc2c6ee6a0cc
Because I changed the expected list of []v1.EnvVar in `TestPodEnvironmentConfigMapVariables` the comparison with `getPodEnvironmentConfigMapVariables` result did not work anymore. But maybe it should be a test on its own together with fetching variables for the secrets and test clearly what overrides what (and also make it clearer in the docs, but I can help there).

dmvolod commented:
> I got it working now. See my gist: https://gist.github.com/FxKu/0b4e0641f1c846a2a0edbc2c6ee6a0cc
> Because I changed the expected list of []v1.EnvVar in `TestPodEnvironmentConfigMapVariables` the comparison with `getPodEnvironmentConfigMapVariables` result did not work anymore. But maybe it should be a test on its own together with fetching variables for the secrets and test clearly what overrides what (and also make it clearer in the docs, but I can help there).

Thanks for test, @FxKu 
Probably, we should move EnvVar logic to the `generateSpiloPodEnvVars` and utilize another tests for that. What do you think about it?

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -939,6 +939,10 @@ func deduplicateEnvVars(input []v1.EnvVar, containerName string, logger *logrus.
 
+	if len(c.Spec.Env) > 0 {
+		configMapPodEnvVarsList = append(configMapPodEnvVarsList, c.Spec.Env...)
+	}
```
Lets to do this only once in `generateStatefulSet`. First get c.Spec.Env because it takes precedence over all other global configs. With the result from `getPodEnvironmentConfigMapVariables` and `getPodEnvironmentSecretVariables` we can do the `isEnvVarPresent` in  `generateStatefulSet`, too. This way, both functions will just do what their names tell us they are supposed to do.

dmvolod replied:
Already did it, what do you think about `generateSpiloPodEnvVars` location?

FxKu replied:
Yes, also a good idea, to move just everything from `generateStatefulSet` to `generateSpiloPodEnvVars`

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -969,6 +975,10 @@ func (c *Cluster) getPodEnvironmentConfigMapVariables() ([]v1.EnvVar, error) {
 
+	if len(c.Spec.Env) > 0 {
+		secretPodEnvVarsList = append(secretPodEnvVarsList, c.Spec.Env...)
+	}
```
see comment above. Maybe we should variables from Secrets should take precedence over variables from config map.

dmvolod replied:
Fix logic, please review.

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -960,7 +964,9 @@ func (c *Cluster) getPodEnvironmentConfigMapVariables() ([]v1.EnvVar, error) {
-		configMapPodEnvVarsList = append(configMapPodEnvVarsList, v1.EnvVar{Name: k, Value: v})
+		if !isEnvVarPresent(configMapPodEnvVarsList, k) {
+			configMapPodEnvVarsList = append(configMapPodEnvVarsList, v1.EnvVar{Name: k, Value: v})
+		}
```
With the result from `getPodEnvironmentConfigMapVariables` we can do the `isEnvVarPresent` in  `generateStatefulSet`, too. This way, the functions will just do what its name tells us it's supposed to do.

dmvolod replied:
Fix logic, please review.

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -1000,20 +1010,31 @@ func (c *Cluster) getPodEnvironmentSecretVariables() ([]v1.EnvVar, error) {
-				SecretKeyRef: &v1.SecretKeySelector{
-					LocalObjectReference: v1.LocalObjectReference{
-						Name: c.OpConfig.PodEnvironmentSecret,
+		if !isEnvVarPresent(secretPodEnvVarsList, k) {
```
Same here. Let's do this in `generateStatefulSet`

dmvolod replied:
Fix logic, please review.

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -1000,20 +1010,31 @@ func (c *Cluster) getPodEnvironmentSecretVariables() ([]v1.EnvVar, error) {
+		}
+	}
+	return false
+}
```
Actually, it might be even better to do `isEnvVarPresent` in [generateSpiloPodEnvVars](https://github.com/zalando/postgres-operator/blob/43e18052c427e255afaa985a046130fc47cf9264/pkg/cluster/k8sres.go#L880), too, instead of appending duplicate variables and resolving it later with `deduplicateEnvVars`.

dmvolod replied:
Fix logic, please review.

FxKu commented on pkg/cluster/k8sres.go:
```diff
@@ -769,7 +769,13 @@ func (c *Cluster) generateSpiloPodEnvVars(
+
+	if len(c.Spec.Env) > 0 {
+		envVars = append(envVars, c.Spec.Env...)
+	}
```
Appending c.Spec.Env should happen after the hard-coded set of variables, because we don't want them to get overridden by custom variables.

dmvolod replied:
Ok, good point, thanks. WIll implement

FxKu commented on pkg/cluster/k8sres_test.go:
```diff
@@ -131,45 +131,52 @@ func TestGenerateSpiloPodEnvVars(t *testing.T) {
+			envVarConstant: "WAL_S3_BUCKET",
+			envVarValue:    "custom-s3-bucket",
+		},
+	}
```
custom variables should not be pre-pended so this addition is also obsolete.

dmvolod replied:
This is approach from the following point
- A custom variable called like a hard-coded variable that can be overridden (like WAL_S3_BUCKET)

FxKu commented on pkg/cluster/k8sres_test.go:
```diff
@@ -207,10 +215,57 @@ func TestGenerateSpiloPodEnvVars(t *testing.T) {
+					},
+				},
+			},
+		},
```
The test should incl. 3 kinds of variables - like in my gist.
1. A custom variable called like a hard-coded variable that cannot be overridden (I uses `KUBERNETES_SCOPE_LABEL`)
2. A custom variable called like a hard-coded variable that can be overridden (like WAL_S3_BUCKET)
3. And a custom variable that's not in the list of hard-coded variables. And there we can test, if a variable from c.Spec.Env overwrites on from environment ConfigMap and/or Secret.

dmvolod replied:
Good point, thanks will implement this options.

dmvolod replied:
Implemented tests, please review.

FxKu commented:
Ok aside from missing unit tests there's one problem that has to be fixed. Right now we only append env vars and rely on `deduplicateEnvVars` to come up with the right set in the end. That is, it will take only the first mention of a variable into account except when the variable starts with `WAL_` or `LOG_`.

This logic will not work anymore now that we have cluster-specific `Env`.
1. It MUST NOT override our hard-coded set of variables from `generateSpiloPodEnvVars` up until this [line](https://github.com/zalando/postgres-operator/blob/43e18052c427e255afaa985a046130fc47cf9264/pkg/cluster/k8sres.go#L868). Only append after that.
2. MUST override the PodEnvironmentConfigMap and/or Secret. So we append before that. But then `WAL_` or `LOG_` settings from PodEnvironmentConfigMap and/or Secret would override `c.Spec.Env` because of the logic in `deduplicateEnvVars`. 

That's why I suggested that we should not simply append to the list of envVars but only do so if the key does not exists yet. So basically what you partly did with `isEnvVarPresent`. But instead of adding this check everywhere, we better write an append function for EnvVar lists that incl. this check.

dmvolod commented:
> That's why I suggested that we should not simply append to the list of envVars but only do so if the key does not exists yet. So basically what you partly did with `isEnvVarPresent`. But instead of adding this check everywhere, we better write an append function for EnvVar lists that incl. this check. Finally, `deduplicateEnvVars` can be removed.

Does it means that we need to utilize new `appendEnvVars` function everywhere inside the `generateSpiloPodEnvVars`?

FxKu commented:
Yes, inside `generateSpiloPodEnvVars`. The flow should be:

1. envVars := []v1.EnvVar{...}
2. append spiloCompathWalPathList
3. append c.Spec.Env
4. fetch from getPodEnvironmentSecretVariables and append
5. fetch from getPodEnvironmentConfigMapVariables and append
6. fetch from generateCloneEnvironment and append
7. fetch from generateStandbyEnvironment and append
8. append WAL_ LOG_ AZURE_ etc. variables

I don't see why anything related to spilo env vars should happen outside of `generateSpiloPodEnvVars`. Noticed that `deduplicateEnvVars` is also used for sidecars. So maybe it can stay then.

dmvolod commented:
> Yes, inside `generateSpiloPodEnvVars`. The flow should be:
> 
> 1. envVars := []v1.EnvVar{...}
> 2. append spiloCompathWalPathList
> 3. append c.Spec.Env
> 4. fetch from getPodEnvironmentSecretVariables and append
> 5. fetch from getPodEnvironmentConfigMapVariables and append
> 6. fetch from generateCloneEnvironment and append
> 7. fetch from generateStandbyEnvironment and append
> 8. append WAL_ LOG_ AZURE_ etc. variables
> 
> I don't see why anything related to spilo env vars should happen outside of `generateSpiloPodEnvVars`. Noticed that `deduplicateEnvVars` is also used for sidecars. So maybe it can stay then.

Thanks for idea, please have a look at the code. Added functions and changed from plain `append` to the new one.

FxKu commented:
:+1:

sdudoladov commented:
:+1:

FxKu commented:
Thanks @dmvolod for your contribution. Good collaboration!

dmvolod commented:
Thanks @FxKu and @sdudoladov for review and ideas.
